zkevm-circuits/src/evm_circuit/util/memory_gadget.rs

```diff
@@ -212,3 +215,168 @@ impl<F: FieldExt, const MAX_QUAD_COST_IN_BYTES: usize>
         Ok((next_memory_size, memory_cost))
     }
 }
+
+///
+#[derive(Clone, Debug)]
+pub(crate) struct BufferGetDataGadget<
+    F,
+    const MAX_BYTES: usize,
+    const ADDR_SIZE_IN_BYTES: usize,
+> {
+    // The bytes that are copied
+    bytes: [Cell<F>; MAX_BYTES],
+    // The selectors that indicate if the bytes contain real data
+    selectors: [Cell<F>; MAX_BYTES],
+    // The LT gadget to check if src_addr is less than src_addr_bound
+    lt_gadget: LtGadget<F, ADDR_SIZE_IN_BYTES>,
+    // The distrance of the offset in the buffer to the buffer_end
+    bound_dist: [Cell<F>; MAX_BYTES],
+    // Check if bound_dist is zero
+    bound_dist_is_zero: [IsZeroGadget<F>; MAX_BYTES],
+}
+
+impl<F: FieldExt, const MAX_BYTES: usize, const ADDR_SIZE_IN_BYTES: usize>
+    BufferGetDataGadget<F, MAX_BYTES, ADDR_SIZE_IN_BYTES>
+{
+    pub(crate) fn construct(
+        cb: &mut ConstraintBuilder<F>,
+        addr_start: &Cell<F>,
+        addr_end: &Cell<F>,
+    ) -> Self {
+        let bytes = array_init(|_| cb.query_byte());
+        let selectors = array_init(|_| cb.query_bool());
+        let bound_dist = array_init(|_| cb.query_cell());
+        let bound_dist_is_zero = array_init(|idx| {
+            IsZeroGadget::construct(cb, bound_dist[idx].expr())
+        });
+        let lt_gadget =
+            LtGadget::construct(cb, addr_start.expr(), addr_end.expr());
+
+        // Define bound_dist[i] = max(addr_end - addr_start - i, 0)
+        // The purpose of bound_dist is to track if the access to src buffer
+        // is out of bound. When bound_dist[i] == 0, it indicates OOB error
+        // and so bytes[i] has to be 0 correspondingly.
+        // Because the bound_dist is decreasing by at most 1 each time, we can
+        // use this property to reduce the use of LtGadget by adding constraints
+        // to the diff between two consecutive bound_dists.
+
+        // Constraints on bound_dist[0]
+        //   bound_dist[0] == 0 || addr_start + bound_dist[0] == addr_end
```

Comment:
Can the situation `addr_start > addr_end` appear ? or don't care about it?

Reply:
In this case `bound_dist[0]` will be 0 as `bound_dist` is defined as `bound_dist[i] = max(addr_end - addr_start - i, 0)`

Reply:
Fix a loophole in the constraint. Could you take a look again? @DreamWuGit

Reply:
good to me !